### ISSUE TITLE
Collaborate with extension proposal

### ISSUE THREAD

**Issue by legendecas** [js-choi/proposal-bind-this#18]
Congratulations on stage 1! 🎉

First of all, I'm strongly supportive of the bind operator things, it is a very good supplementary to the language. Both this proposal and extension proposal has a very similar motivation and similar design, similar semantics on the `::` operator. @hax mentioned that the parallel namespace is not an essential part of the extension proposal and can be removed under the temperature of the committee. And the significant difference between the two proposals is the semantics of binding accessors. I'm wondering if it is better for two proposals to collaborate and find a consensus?

**Comment by nkitku** [comment]
i hope this make sense
```js 
document.querySelectorAll('div')
|> []:filter(%, v => v.clientWidth > 200) // : to unThis 
:: []:filter(v => v.clientHeight > 200) // :: as alternative |> and %, RHS should be function and pass LHS as first param
```

**Comment by js-choi** [comment]
Thanks for the comment, although it doesn’t seem to be related to the extensions proposal (the original topic of this issue). 

I’m going to gently hide both of our comments as off-topic. Please feel free to open a new issue about this idea.

(For what it’s worth, TC39 is trying to minimize the number of operators that get added to the language, so it is unlikely that TC39 would add *three* new operators in addition to the Hack pipe operator. In addition, an unThis operator would lose the [natural word order that a bind operator would restore](https://github.com/tc39/proposal-bind-this/#bind-and-call-are-clunky). So I don’t think we would pursue your idea anyway—thanks anyway, though!)

**Comment by js-choi** [comment]
As the [extensions section][] (and the [in-depth comparison][extensions compare]) say, the concrete
differences between bind-this and extensions are currently:

1. Bind-this has no special variable namespace (but @hax is willing to drop this).
2. Bind-this has no implicit syntactic handling of property accessors.
3. Bind-this has no polymorphic `const ::{ … } from …;` syntax.
4. Bind-this has no polymorphic `…::…:…` syntax.
5. Bind-this has no `Symbol.extension` metaprogramming system.

I’m not yet sure what else @hax is willing to drop, but I plan to keep engaging with him over the next few months, time permitting.

[extensions section]: https://github.com/tc39/proposal-bind-this/#extensions
[extensions compare]: https://github.com/js-choi/proposal-bind-this/blob/main/extensions-comparison.md

**Comment by js-choi** [comment]
An update: I, @tabatkins, and @hax recently wrote articles about the dataflow proposals, including bind-this and Extensions, and TC39 also held two two meetings about the dataflow proposals:

![](https://jschoi.org/21/es-dataflow/map/)

* [My article](https://jschoi.org/21/es-dataflow/)
* [@tabatkins’ article](https://www.xanthir.com/b5Gd0)
* [@hax’s article](https://hackmd.io/yDDJCsS-Sv2AJwo8arAn3w?view)
* [2022-01-26 plenary discussion about holistic dataflow](https://github.com/tc39/notes/blob/main/meetings/2022-01/jan-26.md#holistic-discussion-of-tc39-dataflow-proposals)
* [2022-01-26 post-plenary overflow discussion about holistic dataflow](https://github.com/tc39/incubator-agendas/blob/main/notes/2022/01-27.md)

We will discuss the two proposals further at the next plenary.

**Comment by bergus** [comment]
I must say I *really* dislike the bracketed syntax. Using parenthesis (`rec::(getFn())(arg0)`) should be allowed, but not required for identifiers (`rec::fn(arg0)`) and member expressions (`rec::namespace.fn(arg0)`, `rec::namespace["fn"](arg0)`).

**Comment by hax** [comment]
@bergus As extensions proposal, it also have `rec::method()` syntax so the pair of `rec::method()` and `rec::[method]()` is actually mapping the `rec.method()` and `rec[method]()` syntax.

**Comment by js-choi** [comment]
See also #10. At plenary, the [Committee weakly preferred tight unbracketed receiver-first syntax](https://github.com/babel/proposals/issues/81#issuecomment-1082076382).

**Comment by bergus** [comment]
@hax Are you referring to [this thing](https://github.com/tc39/proposal-extensions/blob/master/docs/design.md#in-place-extension-methodsaccessors)? I don't quite understand why `rec::[method]()` uses bracket syntax - it doesn't appear to evaluate `method` to a property name. I think it would make sense in `rec::namespace:[nameExpr](...args)`, desugaring to `Call(GetOwnProperty(namespace, nameExpr).value, rec, args)` (or `GetExtension` instead of `GetOwnProperty`), but not without a namespace object in which the property name is resolved (since we don't want to resolve it on the receiver).

**Comment by hax** [comment]
@bergus Yes. `v::ns:[expr]` is also possible, or required if we have strong symbol use cases (see https://github.com/tc39/proposal-extensions/issues/11).

Syntax `rec::method / rec::[method]` choose `[]` is just for matching `v.method / v[method]`. But you are right, `rec::[method]` doesn't evaluate `method` to key but evaluate `method` to the method (or accessor) directly.

Of coz, we could consider other brackets syntax if we want to keep `[x]` alway mean evaluating key: `rec::(method)`, `rec::{method}` or even `rec::<method>`, it's just like the current escape hatch syntax of [decorators](https://github.com/tc39/proposal-decorators#syntax) `@(decorator)`.

**Comment by js-choi** [comment]
> Syntax `rec::method / rec::[method]` choose `[]` is just for matching `v.method / v[method]`. But you are right, `rec::[method]` doesn't evaluate `method` to key but evaluate `method` to the method (or accessor) directly.

(For what it’s worth, I actually see some value in an analogy of bound functions acting as pseudo-“methods” of their objects, with the original functions acting as pseudo-“keys”. However, I got some pushback from some others in the Committee about this analogy, so I dropped my pushing it.)

**Comment by hax** [comment]
@js-choi I think the problem may be "bound" is unrelated semantic (of method), and I'm not sure the pushback is because such analogy, my impression is some delegates don't like "free" method. Actually I also feel "free" methods are not common in JS ecosystem, this is why extensions have a special syntax constraints to make it not as "free" as normal functions.

**Comment by ewinslow** [comment]
> I must say I _really_ dislike the bracketed syntax. Using parenthesis (`rec::(getFn())(arg0)`) should be allowed, but not required for identifiers (`rec::fn(arg0)`) and member expressions (`rec::namespace.fn(arg0)`, `rec::namespace["fn"](arg0)`).

If we require parens for member expressions, that to me drops the biggest conflict with the extensions proposal, which includes a way to do extension getters/setters. 

We lose nothing by accepting this requirement. It avoids a whole lot of potential precedence confusion, avoids conflict with the extensions proposal, and doesn't block us from removing parens later if we feel that is justified.

**Comment by js-choi** [comment]
I’m open to requiring parentheses around property access chains in call-this’s RHS, in the spirit of compatibility with the extensions proposal.

And the RHS parentheses can always be made optional for property chains, if the extensions proposal does not eventually progress.

I would also be interested to know how @ljharb feels about this restriction for his use cases.

**Comment by ljharb** [comment]
I suppose I'm comfortable with the parens, but the extensions proposal's current proposed solution (a parallel binding namespace) is an utter nonstarter, and I've yet to see a direction for that proposal that is worth further discussion. If that proposal has something different in mind, then it'd be great to see these two work together.

That said, this one is more basic and simpler, so I think it would be perfectly fine to advance this one as-is and let the extensions proposal adapt around it.

**Comment by hax** [comment]
@js-choi  : I plan to remove "parallel binding namespace", only keep syntax like `obj.ext%%prop` (`%%` as a placeholder) in extensions proposal. I hope it could not only solve @ljharb 's concern, but also make extensions and call-this complementary.
